File: May 2022/May_18.java
class Solution {
    public List<List<Integer>> criticalConnections(int n, List<List<Integer>> connections) {
        graph = new ArrayList[n];
        for(int i = 0; i < n; i++)
            graph[i] = new ArrayList<>();
        for(var v : connections){
            graph[v.get(0)].add(v.get(1));
            graph[v.get(1)].add(v.get(0));
        }
        
        list = new ArrayList<>();
        time = 0;
        dis = new int[n];
        min = new int[n];
        par = new int[n];
        vis = new boolean[n];
        
        helper(0);
        return list;
    }
    
    List<List<Integer>> list;
    List<Integer>[] graph;
    int[] dis, min, par;
    int time;
    boolean[] vis;
    
    private void helper(int node){
        dis[node] = min[node] = time++;
        vis[node] = true;
        
        for(int nbr : graph[node]){
            if(par[node] == nbr)
                continue;
            if(vis[nbr])
                min[node] = Math.min(min[node], dis[nbr]);
            else{
                par[nbr] = node;
                helper(nbr);
                min[node] = Math.min(min[node], min[nbr]);
                if(min[nbr] > dis[node]){
                    List<Integer> al = new ArrayList<>();
                    al.add(node);
                    al.add(nbr);
                    list.add(al);
                }
            }
        }
    }
}

File: May 2022/May_18.cpp
class Solution {
public:
    
    vector<vector<int>> ans;
    int time;
    
    void helper(vector<vector<int>> &graph, vector<int> &par, vector<int> &dis, vector<int> &mini, vector<bool> &vis, int node){
        dis[node] = mini[node] = time++;
        vis[node] = true;
        
        for(auto &nbr : graph[node]){
            if(par[node] == nbr)
                continue;
            if(vis[nbr])
                mini[node] = min(mini[node], dis[nbr]);
            else{
                par[nbr] = node;
                helper(graph, par, dis, mini, vis, nbr);
                mini[node] = min(mini[node], mini[nbr]);
                if(mini[nbr] > dis[node]){
                    vector<int> v(2);
                    v[0] = node;
                    v[1] = nbr;
                    ans.push_back(v);
                }
            }
        }
    }
    
    vector<vector<int>> criticalConnections(int n, vector<vector<int>>& connections) {
        vector<vector<int>> graph(n, vector<int>());
        for(auto &a : connections){
            graph[a[0]].push_back(a[1]);
            graph[a[1]].push_back(a[0]);
        }
        
        vector<int> par(n), dis(n), mini(n);
        vector<bool> vis(n, false);
        time = 0;
        helper(graph, par, dis, mini, vis, 0);
        return ans;
    }
};
